File: NtgeCore.Net/src/Ed25519/Ed25519PublicKey.cs
using System;
using System.Text;
using NtgeCore.Net.Message;
using NtgeCore.Net.X25519;

namespace NtgeCore.Net.Ed25519
{
    public class Ed25519PublicKey : RustObject
    {
        private String? _keyId;
        internal Ed25519PublicKey(IntPtr ptr) : base(ptr)
        {
        }

        public static Ed25519PublicKey Deserialize(string value)
        {
            var ptr = Native.deserializeEd25519PublicKey(Encoding.UTF8.GetBytes(value));
            if (ptr == IntPtr.Zero)
            {
                throw new NtgeException("Can not deserialize private key");
            }
            return new Ed25519PublicKey(ptr);
        }

        public String KeyId 
        {
            get 
            {
                if (_keyId == null)
                {
                    using var handle = Native.publicKeyKeyId(Ptr);
                    _keyId = handle.AsString();
                }
                return _keyId;
            }
        }

        public string Serialize()
        {
            using var result = Native.serializeEd25519PublicKey(Ptr);
            return result.AsString();
        }

        public X25519PublicKey ToX25519()
        {
            return new X25519PublicKey(Native.ed25519PublicKeyToX25519(Ptr));
        }

        public bool Verify(string message, string signature)
        {
            var result = Native.ed25519_public_key_verify(Ptr, Encoding.UTF8.GetBytes(message), Encoding.UTF8.GetBytes(signature));
            return result == 0;
        }

        public override void Dispose()
        {
            Native.destroyEd25519PublicKey(Ptr);
        }

        public override string ToString()
        {
            return Serialize();
        }
    }
}

File: NtgeCore-Android/ntge/src/main/java/com/dimension/ntge/ed25519/Ed25519PublicKey.kt
package com.dimension.ntge.ed25519

import com.dimension.ntge.Ntge
import com.dimension.ntge.x25519.X25519PublicKey

class Ed25519PublicKey internal constructor(
        internal val ptr: Long
) : AutoCloseable {

    companion object {
        fun deserialize(input: String): Ed25519PublicKey {
            return Ed25519PublicKey(Ntge.deserializeEd25519PublicKey(input))
        }
    }

    val keyId by lazy {
        Ntge.publicKeyKeyId(ptr)
    }

    fun serialize(): String {
        return Ntge.serializeEd25519PublicKey(ptr)
    }

    fun toX25519(): X25519PublicKey {
        return X25519PublicKey(Ntge.ed25519PublicKeyToX25519(ptr))
    }

    fun verify(message: ByteArray, signature: ByteArray) = Ntge.ed25519PublicKeyVerify(ptr, message, signature)

    override fun toString(): String {
        return serialize()
    }

    override fun close() {
        Ntge.destroyEd25519PublicKey(ptr)
    }
}

